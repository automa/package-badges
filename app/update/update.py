import logging
import os
from pathlib import Path
from typing import List, Optional

from .ecosystems.cargo import Cargo
from .ecosystems.ecosystem import Ecosystem, Package
from .ecosystems.npm import NPM
from .ecosystems.pypi import PyPI
from .formats.md import MDFormat
from .formats.rst import RSTFormat

ecosystems = [
    NPM(),
    Cargo(),
    PyPI(),
]

formats = [
    MDFormat(),
    RSTFormat(),
]


def update(folder: str) -> List[Package]:
    """
    Find all packages inside the given folder for all supported package managers
    and edit their READMEs to add badges related to package manager.

    Args:
        folder: The root folder to search in
    """
    changed_packages = []

    # Walk through the directory structure
    for root, _, files in os.walk(folder):
        # Check each ecosystem
        for eco in ecosystems:
            if not eco.detect(files):
                continue

            manifest_path = Path(root) / eco.manifest_file
            package_info = eco.parse(root, manifest_path)

            if not package_info:
                continue

            readme_path = find_readme(root, files)

            if readme_path and modify_readme(eco, package_info, readme_path):
                changed_packages.append(package_info)

    return changed_packages


def find_readme(directory: str, files: List[str]) -> Optional[Path]:
    """Find a README file in the given directory."""
    readme_candidates = ["readme.md", "readme.rst"]

    # Use next with generator expression to find the first matching file
    matching_file = next(
        (file for file in files if file.lower() in readme_candidates), None
    )

    return Path(directory) / matching_file if matching_file else None


def modify_readme(ecosystem: Ecosystem, package: Package, readme_path: Path) -> bool:
    is_changed = False

    try:
        with open(readme_path, "r") as f:
            content = f.read()

        # Get badges from the ecosystem
        badges = ecosystem.badges(package)

        # Get the format handler based on file extension
        file_ext = readme_path.suffix.lower()
        format_handler = next(
            (fmt for fmt in formats if fmt.extension == file_ext), None
        )

        if not format_handler:
            raise ValueError(f"Unsupported README format {file_ext}")

        # Use the format handler to modify the content with badges
        new_content = format_handler.modify_content(content, badges)

        # Update the is_changed flag if content has changed
        is_changed = new_content != content

        # Write the updated content back to the file
        with open(readme_path, "w") as f:
            f.write(new_content)

    except Exception as e:
        logging.error(
            f"Error updating README: {e}",
            extra={
                "package": package.name,
                "ecosystem": ecosystem.name,
                "readme": readme_path,
            },
        )

    return is_changed
