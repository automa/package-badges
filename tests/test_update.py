import os
import shutil
from pathlib import Path
from tempfile import TemporaryDirectory
from typing import Dict

import pytest
from syrupy.assertion import SnapshotAssertion

from app.update.update import update


@pytest.fixture
def temp_dir():
    """Create a temporary directory for testing."""
    with TemporaryDirectory() as dir_path:
        yield dir_path


def run_fixture(
    temp_dir: TemporaryDirectory[str], fixture_name: str, snapshot: SnapshotAssertion
) -> None:
    """Run test on a specific fixture."""
    # Setup the fixture
    source = Path(__file__).parent / "fixtures" / fixture_name
    fixture_dir = Path(temp_dir) / fixture_name

    if os.path.isdir(source):
        shutil.copytree(source, fixture_dir)

    # Run the update function
    update(fixture_dir)

    result: Dict[str, str] = {}

    for root, _, files in os.walk(fixture_dir):
        for file in files:
            file_path = Path(root) / file

            with open(file_path, "r") as f:
                content = f.read()
                rel_path = file_path.relative_to(fixture_dir)
                result[str(rel_path)] = content

    assert result == snapshot


def test_empty(temp_dir: TemporaryDirectory[str], snapshot: SnapshotAssertion):
    run_fixture(temp_dir, "empty", snapshot)


def test_no_readme(temp_dir: TemporaryDirectory[str], snapshot: SnapshotAssertion):
    run_fixture(temp_dir, "no_readme", snapshot)


def test_npm(temp_dir: TemporaryDirectory[str], snapshot: SnapshotAssertion):
    run_fixture(temp_dir, "npm", snapshot)


def test_npm_private(temp_dir: TemporaryDirectory[str], snapshot: SnapshotAssertion):
    run_fixture(temp_dir, "npm_private", snapshot)


def test_monorepo(temp_dir: TemporaryDirectory[str], snapshot: SnapshotAssertion):
    run_fixture(temp_dir, "monorepo", snapshot)
