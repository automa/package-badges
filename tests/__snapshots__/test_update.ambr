# serializer version: 1
# name: test_cargo
  dict({
    'Cargo.toml': '''
      [package]
      name    = "test-cargo-crate"
      version = "0.1.0"
  
    ''',
    'README.md': '''
      # Test Cargo Crate
      
      [![Crates.io version](https://img.shields.io/crates/v/test-cargo-crate)](https://crates.io/crates/test-cargo-crate)
      [![Docs.rs documentation](https://img.shields.io/badge/docs.rs-latest-blue)](https://docs.rs/test-cargo-crate)
      [![Crates.io license](https://img.shields.io/crates/l/test-cargo-crate)](https://crates.io/crates/test-cargo-crate)
      
      
      This is a test README for a Cargo crate.
  
    ''',
  })
# ---
# name: test_cargo_private
  dict({
    'Cargo.toml': '''
      [package]
      name    = "test-cargo-crate"
      publish = false
      version = "0.1.0"
  
    ''',
    'README.md': '''
      # Test Cargo Crate
      
      This is a test README for a Cargo crate.
  
    ''',
  })
# ---
# name: test_cargo_publish_both_registries
  dict({
    'Cargo.toml': '''
      [package]
      name    = "test-cargo-crate"
      publish = ["crates-io", "my-registry"]
      version = "0.1.0"
  
    ''',
    'README.md': '''
      # Test Cargo Crate
      
      [![Crates.io version](https://img.shields.io/crates/v/test-cargo-crate)](https://crates.io/crates/test-cargo-crate)
      [![Docs.rs documentation](https://img.shields.io/badge/docs.rs-latest-blue)](https://docs.rs/test-cargo-crate)
      [![Crates.io license](https://img.shields.io/crates/l/test-cargo-crate)](https://crates.io/crates/test-cargo-crate)
      
      
      This is a test README for a Cargo crate.
  
    ''',
  })
# ---
# name: test_cargo_publish_crates_io
  dict({
    'Cargo.toml': '''
      [package]
      name    = "test-cargo-crate"
      publish = ["crates-io"]
      version = "0.1.0"
  
    ''',
    'README.md': '''
      # Test Cargo Crate
      
      [![Crates.io version](https://img.shields.io/crates/v/test-cargo-crate)](https://crates.io/crates/test-cargo-crate)
      [![Docs.rs documentation](https://img.shields.io/badge/docs.rs-latest-blue)](https://docs.rs/test-cargo-crate)
      [![Crates.io license](https://img.shields.io/crates/l/test-cargo-crate)](https://crates.io/crates/test-cargo-crate)
      
      
      This is a test README for a Cargo crate.
  
    ''',
  })
# ---
# name: test_cargo_publish_empty
  dict({
    'Cargo.toml': '''
      [package]
      name    = "test-cargo-crate"
      publish = []
      version = "0.1.0"
  
    ''',
    'README.md': '''
      # Test Cargo Crate
      
      This is a test README for a Cargo crate.
  
    ''',
  })
# ---
# name: test_cargo_publish_registry
  dict({
    'Cargo.toml': '''
      [package]
      name    = "test-cargo-crate"
      publish = ["my-registry"]
      version = "0.1.0"
  
    ''',
    'README.md': '''
      # Test Cargo Crate
      
      This is a test README for a Cargo crate.
  
    ''',
  })
# ---
# name: test_empty
  dict({
    '.gitkeep': '',
  })
# ---
# name: test_md_h3_heading
  dict({
    'README.md': '''
      ### Test NPM Package
      
      [![NPM version](https://img.shields.io/npm/v/test-npm-package)](https://www.npmjs.com/package/test-npm-package)
      [![NPM downloads](https://img.shields.io/npm/dm/test-npm-package)](https://www.npmjs.com/package/test-npm-package)
      [![NPM license](https://img.shields.io/npm/l/test-npm-package)](https://www.npmjs.com/package/test-npm-package)
      
      
      This is a test README for an NPM package.
      
  
    ''',
    'package.json': '''
      {
        "name": "test-npm-package",
        "version": "1.0.0"
      }
  
    ''',
  })
# ---
# name: test_monorepo
  dict({
    'Cargo.toml': '''
      [workspace]
      members = []
  
    ''',
    'crates/cargo/Cargo.toml': '''
      [package]
      name    = "test-cargo-crate"
      version = "0.1.0"
  
    ''',
    'crates/cargo/README.md': '''
      # Test Cargo Crate
      
      [![Crates.io version](https://img.shields.io/crates/v/test-cargo-crate)](https://crates.io/crates/test-cargo-crate)
      [![Docs.rs documentation](https://img.shields.io/badge/docs.rs-latest-blue)](https://docs.rs/test-cargo-crate)
      [![Crates.io license](https://img.shields.io/crates/l/test-cargo-crate)](https://crates.io/crates/test-cargo-crate)
      
      
      This is a test README for a Cargo crate.
  
    ''',
    'package.json': '''
      {
        "name": "monorepo"
      }
  
    ''',
    'packages/README.md': '''
      # monorepo
  
    ''',
    'packages/npm/README.md': '''
      # Test NPM Package
      
      [![NPM version](https://img.shields.io/npm/v/test-npm-package)](https://www.npmjs.com/package/test-npm-package)
      [![NPM downloads](https://img.shields.io/npm/dm/test-npm-package)](https://www.npmjs.com/package/test-npm-package)
      [![NPM license](https://img.shields.io/npm/l/test-npm-package)](https://www.npmjs.com/package/test-npm-package)
      
      
      This is a test README for an NPM package.
  
    ''',
    'packages/npm/package.json': '''
      {
        "name": "test-npm-package",
        "version": "1.0.0"
      }
  
    ''',
  })
# ---
# name: test_no_md_heading
  dict({
    'README.md': '''
      
      [![NPM version](https://img.shields.io/npm/v/test-npm-package)](https://www.npmjs.com/package/test-npm-package)
      [![NPM downloads](https://img.shields.io/npm/dm/test-npm-package)](https://www.npmjs.com/package/test-npm-package)
      [![NPM license](https://img.shields.io/npm/l/test-npm-package)](https://www.npmjs.com/package/test-npm-package)
      
      This is a test README for an NPM package.
  
    ''',
    'package.json': '''
      {
        "name": "test-npm-package",
        "version": "1.0.0"
      }
  
    ''',
  })
# ---
# name: test_no_readme
  dict({
    'package.json': '''
      {
        "name": "test-npm-package",
        "version": "1.0.0"
      }
  
    ''',
  })
# ---
# name: test_npm
  dict({
    'README.md': '''
      # Test NPM Package
      
      [![NPM version](https://img.shields.io/npm/v/test-npm-package)](https://www.npmjs.com/package/test-npm-package)
      [![NPM downloads](https://img.shields.io/npm/dm/test-npm-package)](https://www.npmjs.com/package/test-npm-package)
      [![NPM license](https://img.shields.io/npm/l/test-npm-package)](https://www.npmjs.com/package/test-npm-package)
      
      
      This is a test README for an NPM package.
  
    ''',
    'package.json': '''
      {
        "name": "test-npm-package",
        "version": "1.0.0"
      }
  
    ''',
  })
# ---
# name: test_npm_private
  dict({
    'README.md': '''
      # Test NPM Package
      
      This is a test README for an NPM package.
  
    ''',
    'package.json': '''
      {
        "name": "test-npm-package",
        "version": "1.0.0",
        "private": true
      }
  
    ''',
  })
# ---
